lib/srv/desktop/rdp/rdpclient/client.go

```diff
@@ -0,0 +1,345 @@
+//+build desktop_access_beta
+
+// Package rdpclient implements an RDP client.
+package rdpclient
+
+// Some implementation details that don't belong in the public godoc:
+// This package wraps a Rust library based on https://crates.io/crates/rdp-rs.
+//
+// The Rust library is statically-compiled and called via CGO.
+// The Go code sends and receives the CGO versions of Rust RDP events
+// https://docs.rs/rdp-rs/0.1.0/rdp/core/event/index.html and translates them
+// to the desktop protocol versions.
+//
+// The flow is roughly this:
+//    Go                                Rust
+// ==============================================
+//  rdpclient.New -----------------> connect_rdp
+//                   *connected*
+//
+//            *register output callback*
+//                -----------------> read_rdp_output
+//  handleBitmap  <----------------
+//  handleBitmap  <----------------
+//  handleBitmap  <----------------
+//           *output streaming continues...*
+//
+//              *user input messages*
+//  InputMessage(MouseMove) ------> write_rdp_pointer
+//  InputMessage(MouseButton) ----> write_rdp_pointer
+//  InputMessage(KeyboardButton) -> write_rdp_keyboard
+//            *user input continues...*
+//
+//        *connection closed (client or server side)*
+//    Wait       -----------------> close_rdp
+//
+
+/*
+// Flags to include the static Rust library.
+#cgo linux LDFLAGS: -L${SRCDIR}/target/debug -l:librdp_client.a -lpthread -lcrypto -ldl -lssl -lm
+#cgo darwin LDFLAGS: -framework CoreFoundation -framework Security -L${SRCDIR}/target/debug -lrdp_client -lpthread -lcrypto -ldl -lssl -lm
+#include <librdprs.h>
+*/
+import "C"
+import (
+	"errors"
+	"fmt"
+	"image"
+	"os"
+	"sync"
+	"unsafe"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/srv/desktop/deskproto"
+)
+
+func init() {
+	C.init()
+}
+
+// Options for creating a new Client.
+type Options struct {
+	// Addr is the network address of the RDP server, in the form host:port.
+	Addr string
+	// InputMessage is called to receive a message from the client for the RDP
+	// server. This function should block until there is a message.
+	InputMessage func() (deskproto.Message, error)
+	// OutputMessage is called to send a message from RDP server to the client.
+	OutputMessage func(deskproto.Message) error
+}
+
+func (o Options) validate() error {
+	if o.Addr == "" {
+		return trace.BadParameter("missing Addr in rdpclient.Options")
+	}
+	if o.InputMessage == nil {
+		return trace.BadParameter("missing InputMessage in rdpclient.Options")
+	}
+	if o.OutputMessage == nil {
+		return trace.BadParameter("missing OutputMessage in rdpclient.Options")
+	}
+	return nil
+}
+
+// Client is the RDP client.
+type Client struct {
+	opts Options
+
+	clientWidth, clientHeight uint16
+	username                  string
+	rustClientRef             *C.Client
```

Comment:
Nit: `Ref` suffix seems superfluous, I'd consider just naming this `rustClient` (unless it's intentional).